### PR TITLE
CLDR-13750 Replace custom comparison logic with JDK comparators

### PR DESCRIPTION
**Pull request by hagbard**
This replaces the custom logic for the CollectionUtilities methods with the recently added JDK Comparator(s) functionality. This method probably needs to be double checked against the com.ibm classes to ensure it's equivalent.

See: https://github.com/unicode-org/cldr/blob/master/tools/java/libs/utilities-src.jar

IMPORTANT: This is a clone of https://github.com/unicode-org/cldr/pull/462 which I had to abandon because of conflicting changes caused in the time that this PR has been awaiting review (a combination of the requirement to force squash branches and inability to cleanly merge recent changes).

This is one of the more subtle refactorings since it relies on human understanding of the existing (undocumented, non-trivial) comparison code. Luckily it has only a few uses, and the JDK Comparator API is nice and readable, so the intent of the new code should be clear.

As far as I can tell, the existing comparator code for iterables is equivalent to lexicographical ordering of the “natural” order with the exception that “null” should always be sorted first.
Then the ordering of collections is that, but preceded by a check on the size first.

Note that while the examples in:
tools/java/org/unicode/cldr/tool/GeneratePluralRanges.java
previously passed collections, and thus get a size comparison in the new code, the example in:
tools/java/org/unicode/cldr/util/PreferredAndAllowedHour.java
explicitly passes just the iterator for comparison, bypassing the size check.

Somewhat strangely, the example in:
tools/java/org/unicode/cldr/tool/WritePluralRules.java
does the collection comparison (with size check) and then falls back to a tie-break on the "toString()" of the instances being compared. I've no idea why it's doing this, but I've replicated it.

Note that while I’ve copied this behaviour, I’m not actually sure that in the places it's used you actually expect null at all, so some call sites could possibly be simplified.

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/CLDR-13750
- [x] Updated PR title and link in previous line to include Issue number

### CODE REVIEW

File: tools/java/org/unicode/cldr/tool/WritePluralRules.java [hunk 1]
```diff
@@ -13,15 +15,26 @@
...
 
 public class WritePluralRules {
+    // Ordering by size-of-set first, and then lexicographically, with a final tie-break on the
+    // string representation.
+    // Note: This code was refactored from a legacy utility method (which allowed null) but it's
+    // not actually clear if nulls need to be allowed in this case.
+    private static final Comparator<PluralRules> PLURAL_RULES_COMPARATOR =
+        Comparator.<PluralRules, Integer>comparing(r -> r.getKeywords().size())
+            .thenComparing(PluralRules::getKeywords,
+                Comparators.lexicographical(Comparator.<String>nullsFirst(naturalOrder())))
```
Comment (hagbard) [inline tools/java/org/unicode/cldr/tool/WritePluralRules.java:30]: This and the other "nullsFirst()" call can be removed without breaking anything, and I'd prefer to do that, but you might know why that's not a good idea.

**Comment by hagbard** [comment]
[Comment copied from the abandoned PR #462 for reference]

I did a bit more work on PreferredAndAllowedHour since it was easy to see that nulls could never occur. I improved docs and tightened some of the behaviour around parsing style IDs etc.

One question I have is that why only sometimes is the preferred ID the first one in the allowed set? This is tested for in many tests, but if you add an assertion in the class that it's always true, then code fails in some cases. Is this intended to be a class invariant or not?

**Comment by hagbard** [comment]
I've decided that since nothing breaks when the null-support is removed from the comparator (when running tests and/or the plural rule generation) then it's better to just remove it.

**Comment by macchiati** [comment]
That should be ok. You raise a good issue: filed
https://unicode-org.atlassian.net/browse/CLDR-13852 also.

Mark

**Pull request merged**